src/contracts/deposits-registry/deposits-registry.service.ts

```diff
@@ -0,0 +1,253 @@
+
+    for (
+      let block = firstNotCachedBlock;
+      block <= finalizedBlockNumber;
+      block += DEPOSIT_EVENTS_STEP
+    ) {
+      const chunkStartBlock = block;
+      const chunkToBlock = Math.min(
+        finalizedBlockNumber,
+        block + DEPOSIT_EVENTS_STEP - 1,
+      );
+
+      const chunkEventGroup = await this.fetcher.fetchEventsFallOver(
+        chunkStartBlock,
+        chunkToBlock,
+      );
+
+      await this.sanityChecker.verifyEventsChunk(
+        chunkStartBlock,
+        chunkToBlock,
+        chunkEventGroup.events,
+      );
+
+      // Even if the cache is not valid we can't help but write it down
+      // because the delay in updating the cache will eventually cause
+      // the getAllDepositedEvents method to take a very long time to process, as changes
+      // will be accumulated and not processed.
+      await this.store.insertEventsCacheBatch({
+        headers: {
+          ...initialCache.headers,
+          endBlock: chunkEventGroup.endBlock,
+        },
+        data: chunkEventGroup.events,
+      });
+
+      newEventsCount += chunkEventGroup.events.length;
+
+      const lastEventFromGroup =
+        chunkEventGroup.events[chunkEventGroup.events.length - 1];
+
+      if (lastEventFromGroup) lastIndexedEvent = lastEventFromGroup;
+
+      this.logger.log('Historical events are fetched', {
+        finalizedBlockNumber,
+        startBlock: chunkStartBlock,
+        endBlock: chunkToBlock,
+      });
+    }
+
+    const fetchTimeEnd = performance.now();
+    const fetchTime = Math.ceil(fetchTimeEnd - fetchTimeStart) / 1000;
+
+    const isRootValid = await this.sanityChecker.verifyUpdatedEvents(
+      finalizedBlockHash,
+    );
+
+    // Store the last event from the list of updated events separately
+    // Unfortunately, we cannot validate each event individually upon insertion
+    // because this would require an archival node
+    if (isRootValid && lastIndexedEvent) {
+      await this.store.insertLastValidEvent(lastIndexedEvent);
+    }
+
+    if (!isRootValid) {
+      this.logger.error('Integrity check failed on block', {
+        finalizedBlock,
+        finalizedBlockHash,
+      });
+    }
+
+    this.logger.log('Deposit events cache is updated', {
+      newEventsCount,
+      totalEventsCount: totalEventsCount + newEventsCount,
+      fetchTime,
+    });
+  }
+
+  /**
+   * Returns all deposited events based on cache and fresh data
+   */
+  public async getAllDepositedEvents(
+    blockNumber: number,
+    blockHash: string,
+  ): Promise<VerifiedDepositedEventGroup> {
+    const endBlock = blockNumber;
+    const cachedEvents = await this.getCachedEvents();
+
+    const isCacheValid = this.sanityChecker.verifyCacheBlock(
+      cachedEvents,
+      blockNumber,
+    );
+
+    if (!isCacheValid) {
+      return {
+        events: cachedEvents.data,
+        startBlock: cachedEvents.headers.startBlock,
+        endBlock,
```

Comment:
How endBlock will be used as a result of  this method? look like   const endBlock = blockNumber; here is not correct and endBlock is block from cache

Reply:
changed to
```js
return {
  events: cachedEvents.data,
  startBlock: cachedEvents.headers.startBlock,
  endBlock: cachedEvents.headers.endBlock,
  isValid: false,
};
```
This will be more correct, because in fact we return only what is in the cache + set the validation status (isValid = false).